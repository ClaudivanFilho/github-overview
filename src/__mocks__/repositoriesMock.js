export default [
  {
    id: '1',
    description: 'description',
    url: 'http://github.com/vtex-apps/render-runtime',
    nameWithOwner: 'vtex-apps/render-runtime',
    issues: {
      totalCount: 0,
      nodes: [],
    },
    refs: {
      totalCount: 0,
      nodes: [],
    },
    pullRequests: {
      totalCount: 0,
      nodes: [],
    },
    collaborators: {
      totalCount: 0,
      nodes: [],
    },
  },
]

export const RepositoryMock = {
  'id': 'MDEwOlJlcG9zaXRvcnkxMjkxNDk3Nzk=',
  'description': 'VTEX Minicart app ',
  'createdAt': '2018-04-11T20:23:43Z',
  'url': 'https://github.com/vtex-apps/minicart',
  'nameWithOwner': 'vtex-apps/minicart',
  'collaborators': {
    'totalCount': 55,
    'nodes': [
      {
        'url': 'https://github.com/rsimoens',
        'avatarUrl': 'https://avatars3.githubusercontent.com/u/10401309?v=4',
        'name': 'Raphael Simoens',
      },
      {
        'url': 'https://github.com/vitoriaHeliane',
        'avatarUrl': 'https://avatars3.githubusercontent.com/u/12852518?v=4',
        'name': 'Vitória Heliane',
      },
      {
        'url': 'https://github.com/lcfpadilha',
        'avatarUrl': 'https://avatars1.githubusercontent.com/u/13875462?v=4',
        'name': 'Leonardo Padilha',
      },
      {
        'url': 'https://github.com/mrkmat',
        'avatarUrl': 'https://avatars2.githubusercontent.com/u/14097431?v=4',
        'name': 'Mark Matern',
      },
      {
        'url': 'https://github.com/julopesferr',
        'avatarUrl': 'https://avatars1.githubusercontent.com/u/15784157?v=4',
        'name': 'Juliane Lopes Ferreira',
      },
      {
        'url': 'https://github.com/estacioneto',
        'avatarUrl': 'https://avatars2.githubusercontent.com/u/15948386?v=4',
        'name': 'Estácio Pereira',
      },
      {
        'url': 'https://github.com/luciannojunior',
        'avatarUrl': 'https://avatars3.githubusercontent.com/u/18706156?v=4',
        'name': 'Luciano de Oliveira Júnior',
      },
      {
        'url': 'https://github.com/gabrielgalc',
        'avatarUrl': 'https://avatars2.githubusercontent.com/u/21203990?v=4',
        'name': null,
      },
      {
        'url': 'https://github.com/lowbeer',
        'avatarUrl': 'https://avatars0.githubusercontent.com/u/23402009?v=4',
        'name': null,
      },
      {
        'url': 'https://github.com/luizerasmo',
        'avatarUrl': 'https://avatars0.githubusercontent.com/u/30424413?v=4',
        'name': null,
      },
      {
        'url': 'https://github.com/iduleba',
        'avatarUrl': 'https://avatars0.githubusercontent.com/u/31371069?v=4',
        'name': 'Ian Duleba',
      },
    ],
  },
  'refs': {
    'totalCount': 4,
    'nodes': [
      {
        'name': 'bugfix/grapqlerror',
        'repository': {
          'nameWithOwner': 'vtex-apps/minicart',
        },
        'target': {
          'history': {
            'edges': [
              {
                'node': {
                  'committedDate': '2018-05-12T18:58:01Z',
                  'author': {
                    'name': 'Felipe Sales',
                    'avatarUrl': 'https://avatars1.githubusercontent.com/u/3926634?v=4',
                  },
                },
              },
            ],
          },
        },
      },
      {
        'name': 'fix/duplicate-product',
        'repository': {
          'nameWithOwner': 'vtex-apps/minicart',
        },
        'target': {
          'history': {
            'edges': [
              {
                'node': {
                  'committedDate': '2018-07-03T14:20:39Z',
                  'author': {
                    'name': 'Lucas Cordeiro',
                    'avatarUrl': 'https://avatars3.githubusercontent.com/u/10223856?v=4',
                  },
                },
              },
            ],
          },
        },
      },
      {
        'name': 'master',
        'repository': {
          'nameWithOwner': 'vtex-apps/minicart',
        },
        'target': {
          'history': {
            'edges': [
              {
                'node': {
                  'committedDate': '2018-08-24T15:01:26Z',
                  'author': {
                    'name': 'Bruno Dias',
                    'avatarUrl': 'https://avatars0.githubusercontent.com/u/8000984?v=4',
                  },
                },
              },
            ],
          },
        },
      },
      {
        'name': 'update/quantity-selector',
        'repository': {
          'nameWithOwner': 'vtex-apps/minicart',
        },
        'target': {
          'history': {
            'edges': [
              {
                'node': {
                  'committedDate': '2018-08-29T12:46:12Z',
                  'author': {
                    'name': 'Vitória Heliane',
                    'avatarUrl': 'https://avatars3.githubusercontent.com/u/12852518?v=4',
                  },
                },
              },
            ],
          },
        },
      },
    ],
  },
  'stargazers': {
    'totalCount': 2,
  },
  'issues': {
    'totalCount': 0,
    'nodes': [],
  },
  'pullRequests': {
    'totalCount': 1,
    'nodes': [
      {
        'title': 'Update the Quantity Selector UI and Fix change quantity',
        'url': 'https://github.com/vtex-apps/minicart/pull/37',
        'mergeable': 'MERGEABLE',
        'assignees': {
          'nodes': [
            {
              'avatarUrl': 'https://avatars3.githubusercontent.com/u/12852518?v=4',
              'name': 'Vitória Heliane',
            },
          ],
        },
        'reviews': {
          'nodes': [
            {
              'author': {
                'login': 'ClaudivanFilho',
                'avatarUrl': 'https://avatars2.githubusercontent.com/u/4960686?v=4',
              },
              'state': 'CHANGES_REQUESTED',
            },
            {
              'author': {
                'login': 'salesfelipe',
                'avatarUrl': 'https://avatars1.githubusercontent.com/u/3926634?v=4',
              },
              'state': 'APPROVED',
            },
            {
              'author': {
                'login': 'ClaudivanFilho',
                'avatarUrl': 'https://avatars2.githubusercontent.com/u/4960686?v=4',
              },
              'state': 'CHANGES_REQUESTED',
            },
            {
              'author': {
                'login': 'ClaudivanFilho',
                'avatarUrl': 'https://avatars2.githubusercontent.com/u/4960686?v=4',
              },
              'state': 'APPROVED',
            },
          ],
        },
      },
    ],
  },
}
