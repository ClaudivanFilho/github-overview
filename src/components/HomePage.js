import React, { Component } from 'react'
import PropTypes from 'prop-types'
import RepositoryContext from '../contexts/RepositoryContext'

import BranchList from './BranchList'
import PRList from './PRList'

const LIMIT_DAYS = 30

class HomePage extends Component {
  static propTypes = {
    repositories: PropTypes.arrayOf(PropTypes.object),
  }

  getBranches() {
    let branches = []
    this.props.repositories.map(rep => {
      branches = [...branches, ...rep.refs.nodes]
    })
    return branches
  }

  getPRs() {
    let prs = []
    this.props.repositories.map(rep => {
      prs = [...prs, ...rep.pullRequests.nodes]
    })
    return prs
  }

  getApproves(reviews) {
    const revsLogins = []
    const revsFinal = []
    const revsReversed = reviews.slice().reverse()
    // eslint-disable-next-line
    revsReversed.map(rev => {
      if (!revsLogins.includes(rev.author.login)) {
        revsLogins.push(rev.author.login)
        revsFinal.push(rev)
      }
    })
    return revsFinal.reduce((acc, val) => (
      acc += val.state === 'APPROVED' ? 1 : 0
    ), 0)
  }

  calcDiffDays = (date1, date2) => {
    var oneDay = 24 * 60 * 60 * 1000
    return Math.round(Math.abs((date1.getTime() - date2.getTime()) / (oneDay)))
  }

  branchesFilterBy = (value) => (
    this.calcDiffDays(
      new Date(value.target.history.edges[0].node.committedDate),
      new Date()
    ) > LIMIT_DAYS
  )

  prsFilterBy = (value) => (
    value.mergeable === 'MERGEABLE' && this.getApproves(value.reviews.nodes) > 0
  )

  branchesOrderBy = (val1, val2) => (
    new Date(val1.target.history.edges[0].node.committedDate) >
    new Date(val2.target.history.edges[0].node.committedDate)
  )

  render() {
    return (
      <div className="w-100 flex flex-column flex-row-m pa2 pb4">
        <div className="w-100 w-50-m pa2 pa4-m ma5-m shadow-1">
          <h2 className="mid-gray mv2">Inactive Branches</h2>
          <div className="w-100 br mb2 o-10 ba b--mid-gray"></div>
          <div className="overflow-y-auto f6" style={{ maxHeight: 'calc(100vh - 15rem)' }}>
            <BranchList
              branches={{ nodes: this.getBranches() }}
              seeMore={false}
              filterBy={this.branchesFilterBy}
              orderBy={this.branchesOrderBy}
              showRepository
              dateColor="red"
            />
          </div>
        </div>
        <div className="w-100 w-40-m w-50-l pa2 pa4-m ma5-m mt4 mt0-m shadow-1">
          <h2 className="mid-gray mv2 f4-m mb3-m f3-l mb2-l">Pull Requests (Mergeable)</h2>
          <div className="w-100 br mb2 o-10 ba b--mid-gray"></div>
          <div className="overflow-y-auto f6" style={{ maxHeight: 'calc(100vh - 15rem)' }}>
            <PRList
              pullRequests={{ nodes: this.getPRs() }}
              seeMore={false}
              filterBy={this.prsFilterBy}
              showRepository
            />
          </div>
        </div>
      </div>
    )
  }
}

export default function HomePageWithContext(props) {
  return (
    <RepositoryContext.Consumer>
      {({ repositories }) => <HomePage repositories={repositories} {...props} />}
    </RepositoryContext.Consumer>
  )
}
