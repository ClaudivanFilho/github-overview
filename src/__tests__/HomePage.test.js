import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import HomePage from '../components/HomePage'
import repositoriesMock from '../__mocks__/repositoriesMock'

describe('Test HomePage component', () => {
  let mountedComponent = null

  beforeAll(() => {
    mountedComponent = mount(
      <HomePage repositories={repositoriesMock} />
    )
  })

  it('Renders correctly', () => {
    expect(toJson(mountedComponent)).toMatchSnapshot()
  })

  it('Component should be defined', () => {
    expect(mountedComponent).toBeDefined()
  })
})
