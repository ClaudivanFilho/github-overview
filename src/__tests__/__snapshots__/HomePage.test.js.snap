// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HomePage component Renders correctly 1`] = `
<HomePageWithContext
  repositories={
    Array [
      Object {
        "collaborators": Object {
          "nodes": Array [],
          "totalCount": 0,
        },
        "description": "description",
        "id": "1",
        "issues": Object {
          "nodes": Array [],
          "totalCount": 0,
        },
        "nameWithOwner": "vtex-apps/render-runtime",
        "pullRequests": Object {
          "nodes": Array [],
          "totalCount": 0,
        },
        "refs": Object {
          "nodes": Array [],
          "totalCount": 0,
        },
        "url": "http://github.com/vtex-apps/render-runtime",
      },
    ]
  }
>
  <HomePage
    repositories={
      Array [
        Object {
          "collaborators": Object {
            "nodes": Array [],
            "totalCount": 0,
          },
          "description": "description",
          "id": "1",
          "issues": Object {
            "nodes": Array [],
            "totalCount": 0,
          },
          "nameWithOwner": "vtex-apps/render-runtime",
          "pullRequests": Object {
            "nodes": Array [],
            "totalCount": 0,
          },
          "refs": Object {
            "nodes": Array [],
            "totalCount": 0,
          },
          "url": "http://github.com/vtex-apps/render-runtime",
        },
      ]
    }
  >
    <div
      className="w-100 flex flex-column flex-row-m pa2 pb4"
    >
      <div
        className="w-100 w-50-m pa2 pa4-m ma5-m shadow-1"
      >
        <h2
          className="mid-gray mv2"
        >
          Inactive Branches
        </h2>
        <div
          className="w-100 br mb2 o-10 ba b--mid-gray"
        />
        <div
          className="overflow-y-auto f6"
          style={
            Object {
              "maxHeight": "calc(100vh - 15rem)",
            }
          }
        >
          <BranchList
            branches={
              Object {
                "nodes": Array [],
              }
            }
            dateColor="red"
            filterBy={[Function]}
            orderBy={[Function]}
            seeMore={false}
            showRepository={true}
          />
        </div>
      </div>
      <div
        className="w-100 w-40-m w-50-l pa2 pa4-m ma5-m mt4 mt0-m shadow-1"
      >
        <h2
          className="mid-gray mv2 f4-m mb3-m f3-l mb2-l"
        >
          Pull Requests (Mergeable)
        </h2>
        <div
          className="w-100 br mb2 o-10 ba b--mid-gray"
        />
        <div
          className="overflow-y-auto f6"
          style={
            Object {
              "maxHeight": "calc(100vh - 15rem)",
            }
          }
        >
          <PRList
            filterBy={[Function]}
            pullRequests={
              Object {
                "nodes": Array [],
              }
            }
            seeMore={false}
            showRepository={true}
          />
        </div>
      </div>
    </div>
  </HomePage>
</HomePageWithContext>
`;
